File: crates/traderview-core/src/pyramid.rs
//! Pyramid / scale-in plan calculator.
//!
//! Two related entry-management styles:
//!   * **Pyramid up**: add to winners as price moves in your favor,
//!     each add at a tighter risk band. Used by trend-followers (Dan
//!     Zanger, Mark Minervini).
//!   * **Scale in (averaging down)**: add as price moves AGAINST you
//!     within a planned price ladder. Used by mean-reversion + LEAPs
//!     traders. Very dangerous without a hard total-risk cap.
//!
//! This module computes the size + cost-basis evolution for both. Pure
//! compute. Caller supplies the planned price ladder + tranche sizing.

use crate::models::TradeSide;
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum PlanKind {
    /// Each tranche fires when price moves IN your favor.
    PyramidUp,
    /// Each tranche fires when price moves AGAINST you (averaging down).
    ScaleIn,
}

/// One planned tranche.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tranche {
    /// Price at which this tranche fires.
    pub trigger_price: Decimal,
    /// Number of shares/contracts to add at this tranche.
    pub qty: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanInput {
    pub kind: PlanKind,
    pub side: TradeSide,
    /// First (already-filled) entry.
    pub initial_qty: Decimal,
    pub initial_entry: Decimal,
    pub tranches: Vec<Tranche>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CumulativeState {
    /// Tranche label ("entry", "add 1", "add 2", …)
    pub label: String,
    pub trigger_price: Decimal,
    pub added_qty: Decimal,
    pub total_qty: Decimal,
    pub avg_cost: Decimal,
    /// Cumulative dollar exposure at this state (avg_cost × total_qty).
    pub notional: Decimal,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PlanReport {
    pub states: Vec<CumulativeState>,
    /// Final total qty after all tranches fire.
    pub final_qty: Decimal,
    /// Volume-weighted avg cost across initial + all tranches.
    pub final_avg_cost: Decimal,
    /// Final total notional exposure.
    pub final_notional: Decimal,
    /// True if the plan's tranches violate the kind's direction rule
    /// (e.g., a PyramidUp tranche priced BELOW initial entry for a long).
    pub plan_misordered: bool,
}

pub fn build(input: &PlanInput) -> PlanReport {
    let mut total_qty = input.initial_qty;
    let mut total_cost = input.initial_entry * input.initial_qty;
    let mut states = vec![CumulativeState {
        label: "entry".into(),
        trigger_price: input.initial_entry,
        added_qty: input.initial_qty,
        total_qty,
        avg_cost: input.initial_entry,
        notional: total_cost,
    }];

    let mut misordered = false;
    for (i, t) in input.tranches.iter().enumerate() {
        // Direction check.
        let direction_ok = match (input.kind, input.side) {
            (PlanKind::PyramidUp,  TradeSide::Long)  => t.trigger_price > input.initial_entry,
            (PlanKind::PyramidUp,  TradeSide::Short) => t.trigger_price < input.initial_entry,
            (PlanKind::ScaleIn,    TradeSide::Long)  => t.trigger_price < input.initial_entry,
            (PlanKind::ScaleIn,    TradeSide::Short) => t.trigger_price > input.initial_entry,
        };
        if !direction_ok { misordered = true; }

        total_qty += t.qty;
        total_cost += t.trigger_price * t.qty;
        let avg = if total_qty.is_zero() { Decimal::ZERO } else { total_cost / total_qty };
        states.push(CumulativeState {
            label: format!("add {}", i + 1),
            trigger_price: t.trigger_price,
            added_qty: t.qty,
            total_qty,
            avg_cost: avg,
            notional: avg * total_qty,
        });
    }
    let final_avg = if total_qty.is_zero() { Decimal::ZERO } else { total_cost / total_qty };
    PlanReport {
        final_qty: total_qty,
        final_avg_cost: final_avg,
        final_notional: total_cost,
        states,
        plan_misordered: misordered,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    fn d(s: &str) -> Decimal { Decimal::from_str(s).unwrap() }

    fn pyramid_long() -> PlanInput {
        PlanInput {
            kind: PlanKind::PyramidUp,
            side: TradeSide::Long,
            initial_qty: d("100"),
            initial_entry: d("100"),
            tranches: vec![
                Tranche { trigger_price: d("105"), qty: d("50") },
                Tranche { trigger_price: d("110"), qty: d("25") },
            ],
        }
    }

    #[test]
    fn pyramid_long_increases_qty_and_avg_cost() {
        let r = build(&pyramid_long());
        // After tranche 1: 150 qty @ (100×100 + 50×105)/150 = 101.666...
        // After tranche 2: 175 qty @ (15250 + 25×110)/175 = (15250+2750)/175 = 102.857...
        assert_eq!(r.final_qty, d("175"));
        assert_eq!(r.final_avg_cost.round_dp(4), d("102.8571"));
        assert!(!r.plan_misordered);
    }

    #[test]
    fn scale_in_long_decreases_avg_cost() {
        // Same shape but averaging down.
        let p = PlanInput {
            kind: PlanKind::ScaleIn,
            side: TradeSide::Long,
            initial_qty: d("100"),
            initial_entry: d("100"),
            tranches: vec![
                Tranche { trigger_price: d("95"), qty: d("50") },
                Tranche { trigger_price: d("90"), qty: d("25") },
            ],
        };
        let r = build(&p);
        // After tranche 1: (100×100 + 50×95)/150 = 98.333...
        // After tranche 2: (14750 + 25×90)/175 = 17000/175 = 97.142...
        assert_eq!(r.final_avg_cost.round_dp(4), d("97.1429"));
    }

    #[test]
    fn pyramid_short_uses_lower_triggers() {
        let p = PlanInput {
            kind: PlanKind::PyramidUp,
            side: TradeSide::Short,
            initial_qty: d("100"),
            initial_entry: d("100"),
            tranches: vec![Tranche { trigger_price: d("95"), qty: d("50") }],
        };
        let r = build(&p);
        assert!(!r.plan_misordered);
    }

    #[test]
    fn misorder_flagged_when_pyramid_long_tranche_is_below_entry() {
        let mut p = pyramid_long();
        p.tranches.push(Tranche { trigger_price: d("90"), qty: d("10") });
        let r = build(&p);
        assert!(r.plan_misordered,
            "pyramid-up long add at a lower price violates direction");
    }

    #[test]
    fn misorder_flagged_when_scale_in_long_tranche_is_above_entry() {
        let p = PlanInput {
            kind: PlanKind::ScaleIn,
            side: TradeSide::Long,
            initial_qty: d("100"),
            initial_entry: d("100"),
            tranches: vec![Tranche { trigger_price: d("110"), qty: d("50") }],
        };
        let r = build(&p);
        assert!(r.plan_misordered);
    }

    #[test]
    fn states_capture_running_avg_at_each_tranche() {
        let r = build(&pyramid_long());
        assert_eq!(r.states.len(), 3);   // initial + 2 tranches
        // State 0: entry only — avg cost = entry.
        assert_eq!(r.states[0].avg_cost, d("100"));
        // State 1: after first add.
        assert_eq!(r.states[1].total_qty, d("150"));
        // State 2 = final.
        assert_eq!(r.states[2].total_qty, d("175"));
    }

    #[test]
    fn empty_tranches_yields_initial_only() {
        let p = PlanInput {
            kind: PlanKind::PyramidUp,
            side: TradeSide::Long,
            initial_qty: d("100"),
            initial_entry: d("100"),
            tranches: vec![],
        };
        let r = build(&p);
        assert_eq!(r.states.len(), 1);
        assert_eq!(r.final_qty, d("100"));
        assert_eq!(r.final_avg_cost, d("100"));
    }

    #[test]
    fn zero_qty_tranche_does_not_change_avg_cost() {
        let mut p = pyramid_long();
        p.tranches.push(Tranche { trigger_price: d("200"), qty: Decimal::ZERO });
        let r = build(&p);
        // Last tranche added 0 qty — avg cost shouldn't move.
        let prev_avg = r.states[r.states.len() - 2].avg_cost;
        assert_eq!(r.final_avg_cost, prev_avg);
    }
}

File: frontend/js/views/pyramid.js
// Pyramid / Scale-In view — entry-management planning.
//
// Pyramid-Up: add to winners as price moves in your favor (Minervini /
// Zanger style).
// Scale-In: add as price moves against you within a planned ladder
// (mean-reversion / LEAPs style — risky without a hard total-risk cap).
// Visualizes per-tranche state evolution + avg-cost curve.

import { api } from '../api.js';
import { esc } from '../util.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import {
    parseTrancheBlob, validateInputs, buildBody,
    directionMisordered, decToNum, makeDemoData,
    avgCostSeries, fmtN, fmtInt, fmtUSD,
} from '../_pyramid_inputs.js';

import { t } from '../i18n.js';
import { showToast } from '../toast.js';
let state = {
    kind: 'pyramid_up', side: 'long',
    initialQty: 100, initialEntry: 100,
    trancheText: '',
};

export async function renderPyramid(mount, _appState) {
    const tok = currentViewToken();
    mount.innerHTML = `
        <h1 data-i18n="view.pyramid.h1.pyramid_entry_management" class="view-title">// PYRAMID · ENTRY MANAGEMENT</h1>

        <div class="chart-panel">
            <h2 data-i18n="view.pyramid.h2.strategy">Strategy</h2>
            <div class="inline-form">
                <label><span data-i18n="view.pyramid.label.kind">Kind</span>
                    <select id="py-kind" data-tip="view.pyramid.tip.kind">
                        <option data-i18n="view.pyramid.opt.pyramid_up_add_to_winners" value="pyramid_up" ${state.kind === 'pyramid_up' ? 'selected' : ''}>Pyramid Up (add to winners)</option>
                        <option data-i18n="view.pyramid.opt.scale_in_avg_down_up_against" value="scale_in"   ${state.kind === 'scale_in'   ? 'selected' : ''}>Scale In (avg down / up against)</option>
                    </select></label>
                <label><span data-i18n="view.pyramid.label.side">Side</span>
                    <select id="py-side" data-tip="view.pyramid.tip.side">
                        <option data-i18n="view.pyramid.opt.long" value="long"  ${state.side === 'long'  ? 'selected' : ''}>Long</option>
                        <option data-i18n="view.pyramid.opt.short" value="short" ${state.side === 'short' ? 'selected' : ''}>Short</option>
                    </select></label>
                <label><span data-i18n="view.pyramid.label.initial_qty">Initial qty</span>
                    <input id="py-iq" type="number" step="any" min="0" value="${state.initialQty}" data-tip="view.pyramid.tip.initial_qty"></label>
                <label><span data-i18n="view.pyramid.label.initial_entry">Initial entry $</span>
                    <input id="py-ie" type="number" step="any" min="0" value="${state.initialEntry}" data-tip="view.pyramid.tip.initial_entry"></label>
            </div>
        </div>

        <div class="chart-panel">
            <h2 data-i18n="view.pyramid.h2.tranche_ladder">Tranche ladder</h2>
            <p class="muted" data-i18n="view.pyramid.hint.format">One tranche per line: trigger_price qty. Pyramid-Up: tranches must move INTO profit. Scale-In: tranches must move AGAINST you within your planned ladder. Misordered tranches are flagged by both the local pre-flight and the backend.</p>
            <textarea id="py-tranches" rows="5" placeholder="105 75&#10;110 50&#10;115 25" data-tip="view.pyramid.tip.tranches"></textarea>
            <div class="inline-form">
                <button data-i18n="view.pyramid.btn.demo_pyramid_up_long" id="py-demo-pu-long"  class="secondary" type="button" data-tip="view.pyramid.tip.demo_pu_long">Demo: Pyramid Up Long</button>
                <button data-i18n="view.pyramid.btn.demo_pyramid_up_short" id="py-demo-pu-short" class="secondary" type="button" data-tip="view.pyramid.tip.demo_pu_short">Demo: Pyramid Up Short</button>
                <button data-i18n="view.pyramid.btn.demo_scale_in_long" id="py-demo-si-long"  class="secondary" type="button" data-tip="view.pyramid.tip.demo_si_long">Demo: Scale In Long</button>
                <button data-i18n="view.pyramid.btn.demo_scale_in_short" id="py-demo-si-short" class="secondary" type="button" data-tip="view.pyramid.tip.demo_si_short">Demo: Scale In Short</button>
                <button data-i18n="view.pyramid.btn.build_plan" id="py-run" class="primary" type="button" data-tip="view.pyramid.tip.run" data-shortcut="pyramid_run">Build plan</button>
            </div>
        </div>

        <div id="py-errors" class="boot" style="display:none"></div>
        <div id="py-misorder" class="boot" style="display:none;color:#ffd84a"></div>
        <div id="py-summary" class="cards"></div>

        <div class="chart-panel">
            <h2 data-i18n="view.pyramid.h2.state_evolution">State evolution</h2>
            <div id="py-table"></div>
        </div>

        <div class="chart-panel">
            <h2 data-i18n="view.pyramid.h2.average_cost_curve">Average cost curve</h2>
            <div id="py-chart" style="height:260px"></div>
            <p data-i18n="view.pyramid.hint.cyan_avg_cost_per_share_contract_after_each_tranch" class="muted">Cyan = avg cost per share/contract after each tranche fires.
                Pyramid-Up curves AWAY from initial entry (avg cost rises with adds);
                Scale-In curves TOWARD initial entry (avg cost moves favorably).</p>
        </div>

        <div class="chart-panel">
            <h2 data-i18n="view.pyramid.h2.total_qty_chart">Total position size after each tranche</h2>
            <div id="py-qty-chart" style="width:100%;height:240px"></div>
            <p data-i18n="view.pyramid.hint.total_qty" class="muted small">Cumulative qty after each tranche fires. Pyramid-Up: tapering adds = decreasing slope (smaller-each-add discipline). Scale-In: equal/larger adds = steeper slope (martingale risk).</p>
        </div>

        <div id="py-err" class="boot" style="display:none;color:var(--red)"></div>
    `;

    const loadDemo = (kind, side) => {
        const d = makeDemoData(kind, side);
        document.getElementById('py-kind').value = d.kind;
        document.getElementById('py-side').value = d.side;
        document.getElementById('py-iq').value = d.initial_qty;
        document.getElementById('py-ie').value = d.initial_entry;
        document.getElementById('py-tranches').value =
            d.tranches.map(t => `${t.trigger_price} ${t.qty}`).join('\n');
    };
    document.getElementById('py-demo-pu-long').addEventListener('click',  () => loadDemo('pyramid_up', 'long'));
    document.getElementById('py-demo-pu-short').addEventListener('click', () => loadDemo('pyramid_up', 'short'));
    document.getElementById('py-demo-si-long').addEventListener('click',  () => loadDemo('scale_in',   'long'));
    document.getElementById('py-demo-si-short').addEventListener('click', () => loadDemo('scale_in',   'short'));
    document.getElementById('py-run').addEventListener('click', () => {
        readInputs();
        void compute(tok);
    });
}

function readInputs() {
    state.kind = document.getElementById('py-kind').value;
    state.side = document.getElementById('py-side').value;
    state.initialQty = Number(document.getElementById('py-iq').value);
    state.initialEntry = Number(document.getElementById('py-ie').value);
    state.trancheText = document.getElementById('py-tranches').value;
}

async function compute(tok) {
    hideErr();
    document.getElementById('py-misorder').style.display = 'none';
    const errs = document.getElementById('py-errors');
    errs.style.display = 'none';

    const { tranches, errors } = parseTrancheBlob(state.trancheText);
    if (errors.length) {
        const head = errors.slice(0, 8).map(e =>
            t('common.parse_error_inline', { line: e.line_no, msg: e.message, raw: e.raw.slice(0, 80) })).join('<br>');
        const more = errors.length > 8 ? `<br>${esc(t('common.and_n_more', { n: errors.length - 8 }))}` : '';
        errs.innerHTML = `<strong>${esc(t('common.parse_errors_lead', { n: errors.length }))}</strong><br>${head}${more}`;
        errs.style.display = 'block';
        showToast(t('view.pyramid.toast.parse_error', { n: errors.length }), { level: 'warning' });
        if (tranches.length === 0) return;
    }
    const input = {
        kind: state.kind, side: state.side,
        initial_qty: state.initialQty, initial_entry: state.initialEntry,
        tranches,
    };
    const vErr = validateInputs(input);
    if (vErr) { showErr(vErr); showToast(t('view.pyramid.toast.invalid'), { level: 'warning' }); return; }

    // Local pre-flight: warn before round-trip if misordered.
    if (directionMisordered(state.kind, state.side, state.initialEntry, tranches)) {
        const el = document.getElementById('py-misorder');
        el.innerHTML = t('view.pyramid.preflight.misorder_html', { kind: state.kind.replace('_', '-'), side: state.side });
        el.style.display = 'block';
        showToast(t('view.pyramid.toast.misorder'), { level: 'warning' });
    }

    let report;
    try {
        report = await api.discPyramidPlan(buildBody(input));
    } catch (e) {
        showErr(t("common.error.api", { msg: e.message || e }));
        showToast(t('view.pyramid.toast.api_error'), { level: 'error' });
        return;
    }
    if (!viewIsCurrent(tok)) return;
    renderSummary(report);
    renderTable(report);
    renderChart(report);
    renderTotalQtyChart(report);
    const states = (report.states || []).length;
    showToast(t('view.pyramid.toast.built', { states }), { level: report.plan_misordered ? 'warning' : 'success' });
}

function renderSummary(r) {
    const finalQty   = decToNum(r.final_qty);
    const finalAvg   = decToNum(r.final_avg_cost);
    const finalNotnl = decToNum(r.final_notional);
    document.getElementById('py-summary').innerHTML = [
        card(t('view.pyramid.card.states'),         String((r.states || []).length)),
        card(t('view.pyramid.card.final_qty'),      fmtInt(finalQty)),
        card(t('view.pyramid.card.final_avg_cost'), fmtN(finalAvg)),
        card(t('view.pyramid.card.final_notional'), fmtUSD(finalNotnl)),
        card(t('view.pyramid.card.plan_misordered'), r.plan_misordered ? t('common.yes') : t('common.no'),
            r.plan_misordered ? 'neg' : 'pos'),
    ].join('');
}

function card(label, value, cls = '') {
    return `<div class="card">
        <div class="label">${esc(label)}</div>
        <div class="value ${cls}">${esc(value)}</div>
    </div>`;
}

function renderTable(report) {
    const wrap = document.getElementById('py-table');
    const states = report.states || [];
    if (!states.length) {
        wrap.innerHTML = `<div class="muted" data-i18n="view.pyramid.empty.states">No states.</div>`;
        return;
    }
    wrap.innerHTML = `
        <table class="lq-table">
            <thead><tr>
                <th data-i18n="view.pyramid.th.label">Label</th><th data-i18n="view.pyramid.th.trigger">Trigger $</th><th data-i18n="view.pyramid.th.added_qty">Added qty</th>
                <th data-i18n="view.pyramid.th.total_qty">Total qty</th><th data-i18n="view.pyramid.th.avg_cost">Avg cost</th><th data-i18n="view.pyramid.th.notional">Notional</th>
            </tr></thead>
            <tbody>
                ${states.map(s => `<tr>
                    <td>${esc(s.label)}</td>
                    <td>${esc(fmtN(decToNum(s.trigger_price)))}</td>
                    <td>${esc(fmtInt(decToNum(s.added_qty)))}</td>
                    <td>${esc(fmtInt(decToNum(s.total_qty)))}</td>
                    <td>${esc(fmtN(decToNum(s.avg_cost)))}</td>
                    <td>${esc(fmtUSD(decToNum(s.notional)))}</td>
                </tr>`).join('')}
            </tbody>
        </table>
    `;
}

function renderChart(report) {
    if (!window.uPlot) return;
    const el = document.getElementById('py-chart');
    const { xs, ys } = avgCostSeries(report);
    const triggerYs = (report.states || []).map(s => decToNum(s.trigger_price));
    new window.uPlot({
        title: '', width: el.clientWidth || 600, height: 260,
        scales: { x: {}, y: {} },
        series: [
            { label: t('chart.series.state_') },
            { label: t('chart.series.avg_cost'),    stroke: '#00e5ff', width: 1.5,
              fill: '#00e5ff14', points: { show: true, size: 8 } },
            { label: t('chart.series.trigger_'),   stroke: '#ffd84a', width: 1.0, dash: [4, 4],
              points: { show: true, size: 5 } },
        ],
        axes: [{ stroke: '#aab', size: 28 }, { stroke: '#aab', size: 50 }],
        legend: { show: true },
    }, [xs, ys, triggerYs], el);
}

function renderTotalQtyChart(report) {
    const el = document.getElementById('py-qty-chart');
    if (!el || !window.uPlot) return;
    el.innerHTML = '';
    const states = report.states || [];
    if (states.length < 1) {
        el.innerHTML = `<div class="muted" data-i18n="view.pyramid.empty_qty_chart">${esc(t('view.pyramid.empty_qty_chart'))}</div>`;
        return;
    }
    const xs = states.map((_, i) => i + 1);
    const ys = states.map(s => decToNum(s.total_qty));
    new window.uPlot({
        title: '', width: el.clientWidth || 600, height: 220,
        scales: { x: {}, y: { auto: true } },
        series: [
            { label: t('chart.series.state_') },
            { label: t('view.pyramid.chart.total_qty'),
              stroke: '#b86bff', width: 1.6,
              points: { show: true, size: 10, fill: '#b86bff', stroke: '#b86bff' } },
        ],
        axes: [{ stroke: '#aab', size: 28 }, { stroke: '#aab', size: 50 }],
        legend: { show: true },
    }, [xs, ys], el);
}

function showErr(msg) {
    const el = document.getElementById('py-err');
    el.textContent = msg;
    el.style.display = 'block';
}

function hideErr() { document.getElementById('py-err').style.display = 'none'; }
